### PR TITLE
Add a few entities from Alien Swarm

### PR DESCRIPTION
**Pull request by xen-000**
This PR adds entities and changes some existing ones from Alien Swarm as follows:

- **New entities:**
filter_activator_context: [VDC](https://developer.valvesoftware.com/wiki/Filter_activator_context), now with the ability to filter based on the context value as well.
filter_activator_model: [VDC](https://developer.valvesoftware.com/wiki/Filter_activator_model)
logic_random_outputs: [VDC](https://developer.valvesoftware.com/wiki/Logic_random_outputs), now with 16 outputs instead of 8.

- **Extra functionality for entities:**
env_fade: Added FadeReverse input and its corresponding ReverseFade keyvalue.
math_counter: Added OnChangedFromMin and OnChangedFromMax outputs.
point_teleport: Added TeleportEntity input (teleports the entity specified in the parameter field)

- **Base entity changes:**
Added absvelocity as a keyvalue so entity/player velocity can be directly manipulated.
Added the OnKilled output (fired when an entity receives Kill) and prevent players from being killed.
Added IncrementContext input to allow adding/subtracting values in contexts.

- **Other:**
Reflect the above changes in base.fgd
Added the missing particles needed for the new precipitation effects (see 7abda2a)

### CODE REVIEW

File: mp/src/game/server/pointteleport.cpp [hunk 1]
```diff
@@ -145,3 +147,41 @@ void CPointTeleport::InputTeleport( inputdata_t &inputdata )
...
+            pPlayer->m_Local.m_bDucking = true;
+            pPlayer->m_Local.m_flDucktime = 0.0f;
+            pPlayer->SetViewOffset(VEC_DUCK_VIEW_SCALED(pPlayer));
+            pPlayer->SetCollisionBounds(VEC_DUCK_HULL_MIN, VEC_DUCK_HULL_MAX);
+        }
+    }
+#endif
+
+    pTarget->Teleport(&m_vSaveOrigin, &m_vSaveAngles, NULL);
+}
```
Comment (Gocnak) [inline mp/src/game/server/pointteleport.cpp]: This teleport code looks very identical to the other teleport method, pull the shared bits out into a private internal method and call it from both.

File: mp/src/game/server/logic_random_outputs.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
+//========= Copyright � 1996-2005, Valve Corporation, All rights reserved. ============//
+//
+// Purpose: 
+//
+// $NoKeywords: $
+//
+//=============================================================================//
+#ifndef LOGICRANDOMOUTPUTS_H
+#define LOGICRANDOMOUTPUTS_H
```
Comment (Gocnak) [inline mp/src/game/server/logic_random_outputs.h]: You can replace the ifdef guards with just `#pragma once` 

You can also get rid of the empty comment blocks.

File: mp/src/game/server/logic_random_outputs.h [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+#include "cbase.h"
+#include "entityinput.h"
+#include "entityoutput.h"
+#include "eventqueue.h"
```
Comment (Gocnak) [inline mp/src/game/server/logic_random_outputs.h]: You do not want to include "cbase.h" in a header. Clean up these imports to only inside what you truly need, and forward declare those that you can.

The less includes in a header the better, as it leads to less accidental dependencies (other headers including this header when they needed what this header includes) which also speeds up compiling.

File: mp/game/momentum/materials/particle/bendibeam.vmt [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+	"$vertexcolor"		1
+	"$nocull"			1
+	//"$spriteorientation" "parallel_upright"
+	//"$spriteorigin" "[ 0.50 0.50 ]"
+	"$spriterendermode" 5
+	$overbrightfactor 15
+}
+
+
+
```
Comment (Gocnak) [inline mp/game/momentum/materials/particle/bendibeam.vmt]: Make sure to use spaces in all your text files and added code files, and to remove extra whitespace.

File: mp/src/game/server/EnvFade.cpp [hunk 5]
```diff
@@ -108,6 +113,47 @@ void CEnvFade::InputFade( inputdata_t &inputdata )
...
+        }
+    }
+    else
+    {
+        UTIL_ScreenFadeAll(m_clrRender, ReverseDuration(), HoldTime(), fadeFlags | FFADE_PURGE);
+    }
+
+    m_OnBeginFade.FireOutput(inputdata.pActivator, this);
+}
+
```
Comment (Gocnak) [inline mp/src/game/server/EnvFade.cpp]: This code also looks very similar to the other InputFade, you should pull it out into a shared private method that they both call (use a bool to determine if reverse)

File: mp/game/momentum/particles/particles_manifest.txt [hunk 6]
```diff
@@ -19,6 +19,9 @@ particles_manifest
 	"file"		"particles/water_impact.pcf"
 	"file"		"!particles/muzzleflashes.pcf"
 
+	// New precipitation particles
+	"file"		"particles/rain_fx.pcf"
+
```
Comment (Gocnak) [inline mp/game/momentum/particles/particles_manifest.txt]: Spaces instead of tabs

File: mp/game/bin/base.fgd [hunk 7]
```diff
@@ -4800,6 +4883,7 @@
 
     // Inputs
     input Teleport(void) : "Teleport the target entity."
+    input TeleportEntity(string) : "Teleport the output parameter box entity."
```
Comment (Gocnak) [inline mp/game/bin/base.fgd]: Wouldn't it be an input parameter?

Comment (xen-000) [reply mp/game/bin/base.fgd]: It's the parameter override of an output in hammer.

**Pull request merged**